### PR TITLE
Support categorical data in GPU weighted sketching.

### PR DESCRIPTION
**Pull request by trivialfis**
Related: https://github.com/dmlc/xgboost/issues/6503 .

### CODE REVIEW

File: src/common/hist_util.cu [hunk 1]
```diff
@@ -267,9 +269,10 @@ void ProcessWeightedBatch(int device, const SparsePage& page,
                              batch_it, dummy_is_valid,
                              0, sorted_entries.size(),
                              &cuts_ptr, &column_sizes_scan);
-
-  auto const& h_cuts_ptr = cuts_ptr.ConstHostVector();
   auto d_cuts_ptr = cuts_ptr.DeviceSpan();
+  detail::RemoveDuplicatedCategories(device, info, d_cuts_ptr, &sorted_entries,
```
Comment (hcho3) [inline src/common/hist_util.cu:273]: I don't see a unit test for `RemoveDuplicatedCategories`. Can you write on?

Comment (trivialfis) [reply src/common/hist_util.cu:273]: It's not exposed in header.  Does `TestCategoricalSketch` suffice?

File: tests/cpp/common/test_hist_util.cu [hunk 2]
```diff
@@ -146,7 +157,8 @@ void TestCategoricalSketch(size_t n, size_t num_categories, int32_t num_bins) {
 }
 
 TEST(HistUtil, DeviceSketchCategoricalFeatures) {
-  TestCategoricalSketch(1000, 256, 32);
+  TestCategoricalSketch(1000, 256, 32, false);
```
Comment (hcho3) [inline tests/cpp/common/test_hist_util.cu:160]: How does this test pass, if there are 256 categories and only 32 bins. Then surely `n_uniques` should be 256 and won't be equal to `cuts.TotalBins()`?

File: src/common/hist_util.cu (hunk 1 above)
Comment (hcho3) [reply src/common/hist_util.cu:273]: Maybe. I'd like to understand how `TestCategoricalSketch` is passing. See my comment below.

File: tests/cpp/common/test_hist_util.cu (hunk 2 above)
Comment (trivialfis) [reply tests/cpp/common/test_hist_util.cu:160]: Bin values for each feature are the unique values.

Comment (trivialfis) [reply tests/cpp/common/test_hist_util.cu:160]: There's no sketching if feature type is categorical.

Comment (trivialfis) [reply tests/cpp/common/test_hist_util.cu:160]: That means `num_bin` is ignored for categorical features.

Comment (hcho3) [reply tests/cpp/common/test_hist_util.cu:160]: > That means `num_bin` is ignored for categorical features.

Can you point me to the piece of code that detects whether a given feature is categorical? Is it the following snippet?
https://github.com/dmlc/xgboost/blob/b05073bda5a5b0ab6be6281212037e26ed9c0404/src/common/quantile.cu#L626-L635

Comment (trivialfis) [reply tests/cpp/common/test_hist_util.cu:160]: @hcho3  For the GPU implementation of GK sketching, there are 4 parts:

- hist_utils.cu(h).  This file is the front end of GPU sketch, which extracts sliding windows, remove duplicated categorical values, then push the value into quantile structure.
- Pruning.  This is defined in quantile.cu, the implementation is exactly as described in xgboost paper (more exact than CPU implementation).  For categorical data, the first part has already handle everything, so extra code in pruning for cat data, is to ignore it (don't prune cat features).  Search the use of `IsCat` function.
- Merging.  I haven't don't anything there yet.  It's used by data iter, which is used by `DaskDeviceQuantileDMatrix` and `DeviceQuantileDMatrix`.  Also it's required for distributed setting where we need a global sketch for data.
- Creating the bin values.  That's the code you referenced.

Comment (trivialfis) [reply tests/cpp/common/test_hist_util.cu:160]: Hope that helps.

Comment (hcho3) [reply tests/cpp/common/test_hist_util.cu:160]: Thanks for the detailed reply. Searching for `IsCat` function makes it clear how the sketching algo treats categorical features differently.

> I haven't done anything there yet. It's used by data iter, which is used by DaskDeviceQuantileDMatrix and DeviceQuantileDMatrix.

So fair to say that we can't yet use categorical data in DeviceQuantileDMatrix ?

**Review by hcho3 — APPROVED** [review]

File: tests/cpp/common/test_hist_util.cu (hunk 2 above)
Comment (trivialfis) [reply tests/cpp/common/test_hist_util.cu:160]: @hcho3 DQD is not yet supported.  See items in https://github.com/dmlc/xgboost/issues/6503 .

**Pull request merged**